test/metabase/db/custom_migrations_test.clj

```diff
@@ -1014,3 +1020,445 @@
+        ldap-expected-mapping {"dc=metabase,dc=com" [(+ 1 admin-group-id)]}]
+
+    (testing "Remove admin from group mapping for LDAP, SAML, JWT if they are enabled"
+      (impl/test-migrations ["v48.00-023"] [migrate!]
+        (with-ldap-and-sso-configured ldap-group-mappings sso-group-mappings
+          (migrate!)
+          (is (= ldap-expected-mapping (get-json-setting :ldap-group-mappings)))
+          (is (= sso-expected-mapping (get-json-setting :jwt-group-mappings)))
+          (is (= sso-expected-mapping (get-json-setting :saml-group-mappings))))))
+
+    (testing "remove admin from group mapping for LDAP, SAML, JWT even if they are disabled"
+      (impl/test-migrations ["v48.00-023"] [migrate!]
+        (with-ldap-and-sso-configured ldap-group-mappings sso-group-mappings
+          (mt/with-temporary-raw-setting-values
+            [ldap-enabled "false"
+             saml-enabled "false"
+             jwt-enabled  "false"]
+            (migrate!)
+            (is (= ldap-expected-mapping (get-json-setting :ldap-group-mappings)))
+            (is (= sso-expected-mapping (get-json-setting :jwt-group-mappings)))
+            (is (= sso-expected-mapping (get-json-setting :saml-group-mappings)))))))
+
+    (testing "Don't remove admin group if `ldap-sync-admin-group` is enabled"
+      (impl/test-migrations ["v48.00-023"] [migrate!]
+        (with-ldap-and-sso-configured ldap-group-mappings sso-group-mappings
+          (mt/with-temporary-raw-setting-values
+            [ldap-sync-admin-group "true"]
+            (migrate!)
+            (is (= ldap-group-mappings (get-json-setting :ldap-group-mappings)))))))
+
+    (testing "Running the migration after a previous data-migration still works"
+      (impl/test-migrations ["v48.00-023"] [migrate!]
+        (with-ldap-and-sso-configured ldap-group-mappings sso-group-mappings
+          (#'custom-migrations/migrate-remove-admin-from-group-mapping-if-needed)
+          (migrate!)
+          (is (= ldap-expected-mapping (get-json-setting :ldap-group-mappings)))
+          (is (= sso-expected-mapping (get-json-setting :jwt-group-mappings)))
+          (is (= sso-expected-mapping (get-json-setting :saml-group-mappings))))))))
+
+
+(deftest check-data-migrations-rollback
+  (testing "that rollback causes data_migrations table to reappear with data"
+    (impl/test-migrations ["v48.00-024"] [_]
+      (let [{:keys [db-type ^javax.sql.DataSource
+                    data-source]} mdb.connection/*application-db*
+            ;; approach from schema-migrations-test/migrate-field-database-type-test
+            migrate!              (partial db.setup/migrate! db-type data-source)]
+        ;; 0 because we removed them and normal path won't trigger any
+        (is (<= 0 (:cnt (t2/query-one {:from   [:data_migrations]
+                                       :select [[[:count :id] :cnt]]}))))
+        (migrate! :up)
```

Comment:
this will migrate up all the way to the latest migration, so if we're developing version 52 for example, it'll migrate up to 52.

it's not needed for this test, ideally we should only migrate up to the version that this test is testing against `"v48.00-024"`.

this is to avoid having to update this test in the future, in cases future migrations will change things that this test touch.

to do this you can use the `migrate!` fn binding returned from `impl/test-migrations ["v48.00-024"] [_]`

Reply:
Yeah ideally I'd love to migrate only a single migration down to check that rollback works, but `migrate!` from `impl/test-migrations` does not accept arguments. I've looked into making it accept arguments, but then `run-migrations-in-range!` does not want to go down, and I'm not sure why, so I stopped there. :) I can use `migrate!` to go up though, of course.

Reply:
yea, what people currently do is use the `migrate!` from `test-migrations` macro do the up migrate, then manually calling `db.setup/migrate! :down` and I think it's fine.


FYI in this PR I'm making changes so that `migrate!` from `test-migrations` can do downgrade as well https://github.com/metabase/metabase/pull/34400/files#diff-7841c3c22e27fd535d5d29dc3617f74311ccc74d609b9636e78aae91d41819a5R200-R201